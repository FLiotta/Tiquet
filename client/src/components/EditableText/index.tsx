// Packages
import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import cn from 'classnames';
import useOnclickOutside from "react-cool-onclickoutside";

// Project
import './styles.scss';

interface IProps {
  text: string,
  textClassName?: string,
  inputClassName?: string,
  onSuccess(value, callback): void,
  onCancel?(): void,
  tag: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span' | 'p',
};

const EditableText = ({
  text,
  tag,
  textClassName,
  inputClassName,
  onSuccess,
  onCancel,
  ...rest
}: IProps): JSX.Element => {
  const { handleSubmit, register, errors } = useForm();
  const [editing, setEditing] = useState(false);

  const handleSubmitSuccess = ({ newValue }) => {
    onSuccess(newValue, () => {
      setEditing(false);
    });
  };

  const cancel = () => {
    setEditing(false);
    if (onCancel) {
      onCancel();
    }
  };

  const clickOutsideRef = useOnclickOutside(cancel);

  return (
    <div {...rest} ref={clickOutsideRef} className="editable-text" onClick={() => setEditing(true)}>
      {!editing
        ? React.createElement(tag, { className: cn("editable-text__title", textClassName) }, (
          <>
            {text} <i className="fas fa-pencil-alt editable-text__icon"></i>
          </>
        ))
        : (
          <form onSubmit={handleSubmit(handleSubmitSuccess)} role="form">
            <input
              type="text"
              role="textbox"
              name="newValue"
              className={cn(inputClassName)}
              ref={register({ required: "Required" })}
              defaultValue={text}
            />
          </form>
        )
      }
    </div>
  )
}

export default EditableText;