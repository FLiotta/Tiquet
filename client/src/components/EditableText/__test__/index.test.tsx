// Packages
import * as React from 'react';
import { render } from '@testing-library/react';

// Project
import EditableText from '../index';


describe('<EditableText />', () => {
  it('Matches snapshot', () => {
    const { container } = render(
      <EditableText
        text="test text"
        tag="p"
        onSuccess={jest.fn()}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});