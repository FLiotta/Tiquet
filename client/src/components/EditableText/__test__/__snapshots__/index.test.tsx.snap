// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableText /> Matches snapshot 1`] = `
<div
  class="editable-text"
>
  <p
    class="editable-text__title"
  >
    test text
     
    <i
      class="fas fa-pencil-alt editable-text__icon"
    />
  </p>
</div>
`;
